### ISSUE TITLE
MaxLength not honored on proeprties in derived types

### ISSUE THREAD

**Issue by Tasteful** [aspnet/EntityFramework#4146]
Following Db-context will not respect the `MaxLength` for entity `B`.
```
    public class TestDbContext : DbContext
    {
        public TestDbContext()
        {
        }
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer("Data Source=(local)\\SQL2014; Integrated Security=True; Initial Catalog=EF-TestDb;");
            base.OnConfiguring(optionsBuilder);
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);

            builder.Entity<A>(b =>
            {
                b.Property(p => p.PropA).HasMaxLength(100);
            });

            builder.Entity<B>(b =>
            {
                b.Property(p => p.PropB).HasMaxLength(100);
            });

            builder.Entity<Base>(b =>
            {
                b.HasKey(p => p.Id);
                b.Property(p => p.PropBase).HasMaxLength(100);
            });
        }
    }
    public abstract class Base
    {
        public virtual Guid Id { get; set; }
        public virtual int Index { get; set; }
        public virtual string PropBase { get; set; }
    }
    public class A : Base
    {
        public virtual string PropA { get; set; }
    }
    public class B : Base
    {
        public virtual string PropB { get; set; }
    }
```

Generated schema script (relevant part) where we can see that `MaxLength` not respected for `PropB`.
```
CREATE TABLE [Base] (
    [Id] uniqueidentifier NOT NULL,
    [Discriminator] nvarchar(max) NOT NULL,
    [Index] int NOT NULL,
    [PropBase] nvarchar(100),
    [PropA] nvarchar(100),
    [PropB] nvarchar(max),
    CONSTRAINT [PK_Base] PRIMARY KEY ([Id])
);
```

The generated schema migration (.designer.cs) has `MaxLength`-annotations
```
            modelBuilder.Entity("TestProject.Base", b =>
                {
                    b.ToTable("Base");

                    b.Property<Guid>("Id")
                        .ValueGeneratedOnAdd();

                    b.Property<string>("Discriminator")
                        .IsRequired();

                    b.Property<int>("Index");

                    b.Property<string>("PropBase")
                        .HasAnnotation("MaxLength", 100);

                    b.HasKey("Id");

                    b.HasAnnotation("Relational:DiscriminatorProperty", "Discriminator");

                    b.HasAnnotation("Relational:DiscriminatorValue", "Base");
                });

            modelBuilder.Entity("TestProject.A", b =>
                {
                    b.HasBaseType("TestProject.Base");

                    b.ToTable("A");

                    b.Property<string>("PropA")
                        .HasAnnotation("MaxLength", 100);

                    b.HasAnnotation("Relational:DiscriminatorValue", "A");
                });

            modelBuilder.Entity("TestProject.B", b =>
                {
                    b.HasBaseType("TestProject.Base");

                    b.ToTable("B");

                    b.Property<string>("PropB")
                        .HasAnnotation("MaxLength", 100);

                    b.HasAnnotation("Relational:DiscriminatorValue", "B");
                });
```
Tested with `RC1-final` and `1.0.0-rc2-16319` 

Sample project exists here: [EF-Discriminator-MaxLength.zip](https://github.com/aspnet/EntityFramework/files/68687/EF-Discriminator-MaxLength.zip)

Wrong schema is generated when executing `dnx ef migrations script` on the attached project.

**Comment by rowanmiller** [comment]
Confirmed that this reproduces on the latest bits.

Here is a simpler repro:

```c#
using Microsoft.Data.Entity;

namespace Console.CI
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var context = new BloggingContext())
            {
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();
            }
        }
    }

    class BloggingContext : DbContext
    {
        public DbSet<Blog> Blogs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Sample;Trusted_Connection=True;");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<RssBlog>()
                .Property(b => b.RssFeed)
                .HasMaxLength(450);
        }
    }

    class Blog
    {
        public int BlogId { get; set; }
        public string Url { get; set; }
    }

    class RssBlog : Blog
    {
        public string RssFeed { get; set; }
    }
}
```

Resulting table is:

```SQL
CREATE TABLE [dbo].[Blog](
	[BlogId] [int] IDENTITY(1,1) NOT NULL,
	[Discriminator] [nvarchar](max) NOT NULL,
	[Url] [nvarchar](max) NULL,
	[RssFeed] [nvarchar](max) NULL,
```

**Comment by rowanmiller** [comment]
Confirmed that the max length annotation is correctly assigned during model building, so this is an issue somewhere in migrations/type mapping.

**Comment by rmja** [comment]
@rowanmiller any chance this will make it into rc2? The priority should come from the fact that a wrong schema is created, and cleaning it up afterwards could be a cumbersome task.

**Comment by rowanmiller** [comment]
@rmja this one is not going to be in RC2, it is on the list to fix for RTM though.

**Comment by bricelam** [comment]
Fixed by #5409
